src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java

```diff
@@ -252,84 +259,115 @@ private MqttConnectOptions getConnectionOptions() throws KeyStoreException {
         return connOpts;
     }
 
-    private synchronized void connectAndSubscribe() {
-        if (connectFuture != null) {
-            connectFuture.cancel(true);
+    private void reconnectAndSubscribeAsync() {
+        synchronized (connectLock) {
+            if (connectFuture != null) {
+                connectFuture.cancel(true);
+            }
+            connectFuture = executorService.submit(() -> {
+                if (mqttClientInternal.isConnected()) {
+                    try {
+                        mqttClientInternal.disconnect();
+                    } catch (MqttException e) {
+                        LOGGER.atError().setCause(e)
+                                .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
+                                .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
+                                .log("Failed to disconnect");
+                        return;
+                    }
+                    try {
+                        connect();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        return;
+                    }
+                }
+                updateSubscriptionsInternal();
+            });
         }
+    }
 
-        LOGGER.atInfo()
-                .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
-                .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
-                .log("Connecting to broker");
-
-        connectFuture = executorService.submit(this::reconnectAndResubscribe);
+    private void connectAndSubscribeAsync() {
+        synchronized (connectLock) {
+            if (connectFuture != null) {
+                connectFuture.cancel(true);
+            }
+            connectFuture = executorService.submit(() -> {
+                try {
+                    connect();
```

Comment:
is this required? Doesn't the client automatically reconnect when needed?

Reply:
there's an option in paho for auto reconnect but we don't use that afaik

Reply:
Why not? Should we?